### PR TITLE
Change architecture & add start to draft-06 support

### PR DESCRIPTION
**Pull request by erayd**
## What
 * Rearchitect to split rules into separate files with a common base
 * Change the API to improve ease of use
 * Initial work on draft-06 support

## Why
 * Ease of maintenance
 * Draft-06 is a required core feature

### CODE REVIEW

**Comment by erayd** [comment]
FYI, this will probably fail travis - there is a bug in json-schema which causes one of the tests to fail.

**Comment by shmax** [comment]
It's midnight here, and in all honesty these days my brain is no good after 6:00 pm, but I had a quick peek and so far the parts I understand look good. Unbelievable amount of work you put into this! Will go through it in more detail tomorrow.

**Comment by erayd** [comment]
Thanks - I look forward to seeing your review then :-).

File: composer.json [hunk 1]
```diff
@@ -10,7 +10,8 @@
     "require-dev": {
         "phpunit/phpunit": "^4.8",
         "squizlabs/php_codesniffer": "^2.7",
-        "satooshi/php-coveralls": "^1.0.1"
+        "satooshi/php-coveralls": "^1.0.1",
+        "justinrainbow/json-schema": "^5.2"
```
Comment (shmax) [inline composer.json]: Is this just for debugging purposes? This package doesn't have some kind of inverse dependency on the other package, does it?

File: rules/base.json [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+            },
+            "propertyNames": {
+                "allow-types": ["boolean", "object"],
+                "for-types": ["object"],
+                "as-schema": true
+            },
+            "type": {
+                "allow-types": ["string", "array"]
+            },
+            "const": {},
```
Comment (shmax) [inline rules/base.json]: Nothin'?

File: src/RuleInfo.php [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+    /**
+     * Get a rule value
+     *
+     * @param string $rule Rule name
+     * @return boolean
+     */
+    public function __get($rule)
+    {
+        $instance = $this;
```
Comment (shmax) [inline src/RuleInfo.php]: unused?

File: composer.json (hunk 1 above)
Comment (erayd) [inline composer.json]: Exactly - it's used for some of the unit tests, but nothing else, and is not a runtime dependency (it/s in the require-dev section).

File: rules/base.json (hunk 2 above)
Comment (erayd) [inline rules/base.json]: This is correct - const can be of any type, for any type. I could have actually omitted it completely, but decided to add it here as an empty object to emphasise that it hadn't been forgotten.

File: src/RuleInfo.php (hunk 3 above)
Comment (erayd) [inline src/RuleInfo.php]: Good catch - this was a workaround for PHP-5.3's shitty closure support, but is no longer necessary as the closure that previously lived here has gone. I'll nuke it now.

**Comment by erayd** [comment]
Thanks for reviewing this. Good to merge?

**Pull request merged**